### PR TITLE
Upgrade to wasmer 1.0.1 and add in-memory cache

### PR DESCRIPTION
**Pull request by thebevrishot**
## Implementation details
- Upgrade wasmer to 1.0.1
- Implement in-memory caching for compiled module but we would not manage it here, we allow API callers (e.g. `go-owasm`) and pass it with code.

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [x] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: src/core/vm.rs [hunk 1]
```diff
@@ -22,31 +26,112 @@ pub trait Env {
...
-        if self.gas_used > self.gas_limit {
+        if self.out_of_gas() {
             Err(Error::OutOfGasError)
         } else {
             Ok(())
         }
     }
+
+    pub fn out_of_gas(&self) -> bool {
+        self.gas_used >= self.gas_limit
```
Comment (taobun) [inline src/core/vm.rs]: Should be `>` ?

File: src/core/cache.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+impl InMemoryCache {
+    /// Creates a new cache with the given size (in bytes)
+    pub fn new(size: Size) -> Self {
+        let max_entries = size.0 / ESTIMATED_MODULE_SIZE.0;
+        InMemoryCache { modules: CLruCache::new(max_entries) }
+    }
+
+    pub fn store(&mut self, checksum: &Checksum, module: &Module) -> Result<(), Error> {
```
Comment (taobun) [inline src/core/cache.rs]: module should be `Module` (store should be the ownership of this clone inside function seems to be the wrong pattern)

Comment (taobun) [inline src/core/cache.rs]: I think our code base didn't use VmResult coding style, so Don't return Result if it don't want to return error

File: src/core/cache.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        self.modules.put(*checksum, module.clone());
+        Ok(())
+    }
+
+    /// Looks up a module in the cache and creates a new module
+    pub fn load(&mut self, checksum: &Checksum) -> Result<Option<Module>, Error> {
```
Comment (taobun) [inline src/core/cache.rs]: Like this it should just return Option<Module> until we bring the Result coding

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by Benzbeeb — APPROVED** [review]

**Pull request merged**